File: csharp/src/Scon.Core/Formatter.cs
namespace RealmLabs.Scon;

internal static class Formatter
{
    public static string FormatValue(SconValue value) => value is SconObject obj ? FormatObjectBody(obj, 0) + "\n" : throw new SconException(ErrorCode.InvalidRootType, "SCON document root must be an object");
    private static string FormatObjectBody(SconObject obj, int indent) { var outText = new StringBuilder(); foreach (var (key, value) in obj) outText.Append(' ', indent).Append(FormatKey(key)).Append(" = ").Append(FormatScon(value, indent)).Append('\n'); return outText.ToString(); }
    private static string FormatScon(SconValue value, int indent) => value switch
    {
        SconNull => "null",
        SconBool b => b.Value ? "true" : "false",
        SconString s => Quote(s.Value, true),
        SconNumberValue n => n.Number.ToSconString(),
        SconArray a => a.Count == 0 ? "[]" : "[\n" + string.Concat(a.Select(item => new string(' ', indent + 2) + FormatScon(item, indent + 2) + ",\n")) + new string(' ', indent) + "]",
        SconObject o => o.Count == 0 ? "{}" : "{\n" + FormatObjectBody(o, indent + 2) + new string(' ', indent) + "}",
        _ => throw new SconException(ErrorCode.Serde, "unsupported SCON value"),
    };
    private static string FormatKey(string key) => Regex.IsMatch(key, "^[A-Za-z_][A-Za-z0-9_-]*$") ? key : Quote(key, false);
    private static string Quote(string value, bool escapeInterpolation)
    {
        var outText = new StringBuilder("\"");
        foreach (var ch in value) outText.Append(ch switch { '"' => "\\\"", '\\' => "\\\\", '\n' => "\\n", '\r' => "\\r", '\t' => "\\t", '\b' => "\\b", '\f' => "\\f", '$' when escapeInterpolation => "\\$", _ when char.IsControl(ch) => $"\\u{(int)ch:X4}", _ => ch.ToString() });
        return outText.Append('"').ToString();
    }
}

File: java/scon-core/src/main/java/io/github/realmlabs/scon/Formatter.java
package io.github.realmlabs.scon;

import java.util.Map;

final class Formatter {
    private Formatter() {}

    static String formatValue(SconValue value) {
        if (!(value instanceof SconObject object)) {
            throw new SconException(ErrorCode.InvalidRootType, "SCON document root must be an object");
        }
        return formatObjectBody(object, 0) + "\n";
    }

    private static String formatObjectBody(SconObject object, int indent) {
        var out = new StringBuilder();
        for (Map.Entry<String, SconValue> entry : object.entrySet()) {
            out.append(" ".repeat(indent))
                .append(formatKey(entry.getKey()))
                .append(" = ")
                .append(formatScon(entry.getValue(), indent))
                .append('\n');
        }
        return out.toString();
    }

    private static String formatScon(SconValue value, int indent) {
        if (value == SconNull.INSTANCE) return "null";
        if (value instanceof SconBool bool) return bool.value() ? "true" : "false";
        if (value instanceof SconString string) return quote(string.value(), true);
        if (value instanceof SconNumber number) return number.toSconString();
        if (value instanceof SconArray array) {
            if (array.isEmpty()) return "[]";
            var out = new StringBuilder("[\n");
            for (SconValue item : array) {
                out.append(" ".repeat(indent + 2)).append(formatScon(item, indent + 2)).append(",\n");
            }
            return out.append(" ".repeat(indent)).append(']').toString();
        }
        if (value instanceof SconObject object) {
            if (object.isEmpty()) return "{}";
            return "{\n" + formatObjectBody(object, indent + 2) + " ".repeat(indent) + "}";
        }
        throw new SconException(ErrorCode.Serde, "unsupported SCON value");
    }

    private static String formatKey(String key) {
        return key.matches("[A-Za-z_][A-Za-z0-9_-]*") ? key : quote(key, false);
    }

    private static String quote(String value, boolean escapeInterpolation) {
        var out = new StringBuilder("\"");
        for (int i = 0; i < value.length(); i++) {
            char ch = value.charAt(i);
            switch (ch) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                case '\b' -> out.append("\\b");
                case '\f' -> out.append("\\f");
                case '$' -> out.append(escapeInterpolation ? "\\$" : "$");
                default -> {
                    if (Character.isISOControl(ch)) out.append(String.format("\\u%04X", (int) ch));
                    else out.append(ch);
                }
            }
        }
        return out.append('"').toString();
    }
}
